src/net/client/cache.rs

```diff
@@ -0,0 +1,1200 @@
+    /// (five minutes) and the default is 30 seconds.
+    pub fn set_transport_failure_duration(&mut self, value: Duration) {
+        self.transport_failure_duration =
+            TRANSPORT_FAILURE_DURATION.limit(value)
+    }
+
+    /// Set the maximum time to cache results other than NOERROR or NXDOMAIN.
+    ///
+    /// The value has to be at least one second, at most 300 seconds
+    /// (five minutes) and the default is 30 seconds.
+    pub fn set_misc_error_duration(&mut self, value: Duration) {
+        self.misc_error_duration = MISC_ERROR_DURATION.limit(value)
+    }
+
+    /// Set the maximum time to cache NXDOMAIN results.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most 86,400
+    /// seconds (one day) and the default is 3,600 seconds (one hour).
+    pub fn set_max_nxdomain_validity(&mut self, value: Duration) {
+        self.max_nxdomain_validity = MAX_NXDOMAIN_VALIDITY.limit(value)
+    }
+
+    /// Set the maximum time to cache NODATA results.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most 86,400
+    /// seconds (one day) and the default is 3,600 seconds (one hour).
+    pub fn set_max_nodata_validity(&mut self, value: Duration) {
+        self.max_nodata_validity = MAX_NODATA_VALIDITY.limit(value)
+    }
+
+    /// Set the maximum time to cache delegations.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most
+    /// 1,000,000,000 seconds and the default is 1,000,000 seconds.
+    pub fn set_max_delegation_validity(&mut self, value: Duration) {
+        self.max_delegation_validity = MAX_DELEGATION_VALIDITY.limit(value)
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            max_cache_entries: MAX_CACHE_ENTRIES.default(),
+            max_validity: MAX_VALIDITY.default(),
+            transport_failure_duration: TRANSPORT_FAILURE_DURATION.default(),
+            misc_error_duration: MISC_ERROR_DURATION.default(),
+            max_nxdomain_validity: MAX_NXDOMAIN_VALIDITY.default(),
+            max_nodata_validity: MAX_NODATA_VALIDITY.default(),
+            max_delegation_validity: MAX_DELEGATION_VALIDITY.default(),
+        }
+    }
+}
+
+//------------ Connection -----------------------------------------------------
+
+#[derive(Clone)]
+/// A connection that caches responses from an upstream connection.
+pub struct Connection<Upstream, C: Clock + Send + Sync = SystemClock> {
+    /// Upstream transport to use for requests.
+    upstream: Upstream,
+
+    /// The cache for this connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of this connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<Upstream> Connection<Upstream> {
+    /// Create a new connection with default configuration parameters.
+    pub fn new(upstream: Upstream) -> Self {
```

Comment:
In the generated RustDoc the lack of trait bounds on `Upstream` makes it very unclear what the caller should pass in as an argument to `nw()`. One way to make this clearer would be to add a `Upstram: SendRequest<CR>` bound on `new()`, or maybe even better at the `struct` level where it will appear at the top of hte generated docs.

Reply:
Generally, bounds are kept to the minimum necessary set of traits. So this should be in the struct’s docstring.

Reply:
But isn't it a minimum requirement anyway in this case to be able to actually use the struct as intended?

Reply:
Minimum requirement for the particular block. For the struct definition, you only need `Upstream` to be `Sized` (which is implied).

Reply:
Given that trait bounds trickle through when types are used, limiting them to what’s needed keeps where clauses short. Similarly for just all the impl blocks for the types themselves.

Reply:
I find the usability more important than the length of trait bounds. Either way this needs to be clear to a user, whether through a RustDoc comment, a helpful compilation error or a less helpful compilation error.